main/src/ca/uwaterloo/flix/api/lsp/Visitor.scala

```diff
@@ -0,0 +1,937 @@
+    * @param pos  the position that must be within the AST node `SourceLocation` for the node to be accepted.
+    */
+  case class InsideAcceptor(uri: String, pos: Position) extends Acceptor {
+    def accept(loc: SourceLocation): Boolean = inside(uri, pos)(loc)
+  }
+
+  /**
+    * Visits the root AST node and recursively visits
+    * all children that are accepted by the acceptor,
+    * "consuming" each node visited (including the root).
+    *
+    * By "consuming" what is meant is that the node is used
+    * as the input to a function of the consumer associated with
+    * the AST node type. Such consumer funcitons have no output,
+    * but can have effects. For instance, the consumer can be defined such that each
+    * expression containing a specific variable is collected in a list, via
+    * mutation a variable.
+    *
+    * Note that if a node is not accepted, none of the nodes in it's
+    * subtree will be visited either. For instance, if a `def` is not
+    * accepted, none of the expressions in the body will be visited eiter,
+    * even if they would otherwise be accepted by the acceptor.
+    *
+    * @param root      The AST root node.
+    * @param consumer  A consumer which defines what to do when visiting different types of AST nodes.
+    * @param acceptor  An acceptor which defines the criteria for whether an AST node should be visited.
+    */
+  def visitRoot(root: Root, consumer: Consumer, acceptor: Acceptor): Unit = {
+
+    implicit val c: Consumer = consumer
+    implicit val a: Acceptor = acceptor
+
+    root.defs.values.foreach(visitDef)
+
+    root.effects.values.foreach(visitEffect)
+
+    root.enums.values.foreach(visitEnum)
+
+    root.instances.values.flatten.foreach(visitInstance)
+
+    root.sigs.values.foreach(visitSig)
+
+    root.structs.values.foreach(visitStruct)
+
+    root.traits.values.foreach(visitTrait)
+
+    root.typeAliases.values.foreach(visitTypeAlias)
+  }
+
+  /**
+    * Returns `true` if the position `pos` within the file given by path `uri` is contained within the `SourceLocation` `loc`.
+    * Returns `false` otherwise.
+    *
+    * @param uri  the path of the file that `pos` is within.
+    * @param pos  the position that we want to know whether is within `loc`.
+    * @param loc  the `SourceLocation` that want to know if `pos` is within.
+    * @return `true` if `pos` in file at path `uri` is within `loc`. `false` otherwise.
+    */
+  def inside(uri: String, pos: Position)(loc: SourceLocation): Boolean = {
+    val sameSource = uri == loc.source.name
+    if (!sameSource) { return false }
+
+    val afterStart = loc.beginLine < pos.line ||
+                     loc.beginLine == pos.line && loc.beginCol <= pos.character
+    if (!afterStart) { return false }
+
+
+    val beforeEnd = pos.line < loc.endLine ||
+                    pos.line == loc.beginLine && pos.character < loc.endCol
+    if (!beforeEnd) { return false }
+
+    true
```

Comment:
Much better- thanks!

Reply:
Do you have 100% test coverage?

Reply:
No, I did not. I have added more extensive testing (maybe too extensive tbh), and now there should be coverage on the entire inside function. These new tests also did shed light on a dumb mistake I made in `inside` that went over my head